pkg/trigger/discovery_handler_test.go

```diff
@@ -0,0 +1,282 @@
+		return
+	}
+	trigger.heartbeat = metav1.Now()
+	discoveredResource := newAPIResource()
+	trigger.processDiscoveryResource(discoveredResource)
+	actions := client.Actions()
+	verifyCleanupAndLaunch(t, actions[3:7])
+
+	c, ok := actions[8].(core.CreateAction)
+	if !ok {
+		t.Fatalf("expected create action")
+	}
+	r := schema.GroupVersionResource{Group: "migration.k8s.io", Version: "v1alpha1", Resource: "storagestates"}
+	if c.GetResource() != r {
+		t.Fatalf("unexpected resource %v", c.GetResource())
+	}
+
+	verifyStorageStateUpdate(t, actions[9], trigger.heartbeat, discoveredResource.StorageVersionHash, []string{v1alpha1.Unknown})
+}
+
+func TestProcessDiscoveryResourceStaleState(t *testing.T) {
+	client := fake.NewSimpleClientset(newMigrationList(), staleStorageState())
+	trigger := NewMigrationTrigger(client)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go trigger.migrationInformer.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, trigger.migrationInformer.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Unable to sync caches"))
+		return
+	}
+	trigger.heartbeat = metav1.Now()
+	discoveredResource := newAPIResource()
+	trigger.processDiscoveryResource(discoveredResource)
+
+	actions := client.Actions()
+	d, ok := actions[3].(core.DeleteAction)
+	if !ok {
+		t.Fatalf("expected delete action")
+	}
+	r := schema.GroupVersionResource{Group: "migration.k8s.io", Version: "v1alpha1", Resource: "storagestates"}
+	if d.GetResource() != r {
+		t.Fatalf("unexpected resource %v", d.GetResource())
+	}
+	if !strings.Contains(d.GetName(), staleStorageState().Name) {
+		t.Fatalf("unexpected name %s", d.GetName())
+	}
+
+	verifyCleanupAndLaunch(t, actions[4:8])
+
+	c, ok := actions[9].(core.CreateAction)
+	if !ok {
+		t.Fatalf("expected create action")
+	}
+	r = schema.GroupVersionResource{Group: "migration.k8s.io", Version: "v1alpha1", Resource: "storagestates"}
+	if c.GetResource() != r {
+		t.Fatalf("unexpected resource %v", c.GetResource())
+	}
+
+	verifyStorageStateUpdate(t, actions[10], trigger.heartbeat, discoveredResource.StorageVersionHash, []string{v1alpha1.Unknown})
+}
+
+func TestProcessDiscoveryResourceStorageVersionChanged(t *testing.T) {
+	client := fake.NewSimpleClientset(newMigrationList(), freshStorageStateWithOldHash())
+	trigger := NewMigrationTrigger(client)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go trigger.migrationInformer.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, trigger.migrationInformer.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Unable to sync caches"))
+		return
+	}
+	trigger.heartbeat = metav1.Now()
+	discoveredResource := newAPIResource()
+	trigger.processDiscoveryResource(discoveredResource)
+
+	actions := client.Actions()
+	verifyCleanupAndLaunch(t, actions[3:7])
+	verifyStorageStateUpdate(t, actions[8], trigger.heartbeat, discoveredResource.StorageVersionHash, []string{"oldhash", "newhash"})
+}
+
+func TestProcessDiscoveryResourceNoChange(t *testing.T) {
```

Comment:
this function looks to be the same as the function above

Reply:
Thanks. Updated.